### ISSUE TITLE
pycairo build failure

### ISSUE THREAD

**Issue by comotion** [wingtk/gvsbuild#899]
Hi,
trying to build pygobject here on Windows 10 with Anaconda3 latest, with this command:
`gvsbuild.exe build --platform x86 --enable-gi --py-wheel pycairo pygobject` 

Most things will build but pycairo fails with:

```(tar) Exporting pycairo
Building project pycairo (1.23.0)
The Meson build system
Version: 1.0.1
Source dir: C:\gtk-build\build\Win32\release\pycairo
Build dir: C:\gtk-build\build\Win32\release\pycairo\_gvsbuild-meson
Build type: native build
Project name: pycairo
Project version: 1.23.0
C compiler for the host machine: cl (msvc 19.35.32215 "Microsoft (R) C/C++ Optimizing Compiler Version 19.35.32215 for x86")
C linker for the host machine: link link 14.35.32215.0
Host machine cpu family: x86
Host machine cpu: x86
Program nonexistingprogram found: NO

meson.build:15:0: ERROR: python3 not found

A full log can be found at C:\gtk-build\build\Win32\release\pycairo\_gvsbuild-meson\meson-logs\meson-log.txt
```

The pycairo build seems to fail to find the anaconda python. Installing another python3 with `choco install python39 -x86` fixes it.

**Comment by danyeaw** [comment]
Hi @comotion, I don't think there is anything we can do about this. Meson supports the recommended way to install Python on Windows using the official installers. If you would like to have better support for Anaconda, that should probably be requested upstream to the meson project.

**Comment by tjwilli58** [comment]
Hi @danyeaw , @comotion ,  I also got this error trying to rebuild my gstreamer build. I'm not using Anaconda for this. I'm using the regular Windows Python installer. I had gstreamer, gst-python, gst-plugins-bad, etc built back in December. (issue https://github.com/wingtk/gvsbuild/issues/796).

I think the issue _may_ be me upgrading from Python 3.9 to 3.11. I'm going to go back to 3.9 and see what happens.

**Comment by danyeaw** [comment]
Interesting, I am using Python 3.11 and our CI is as well.

**Comment by tjwilli58** [comment]
Using python 3.11, I get the same error as @comotion. Started with a clean build-dir.
```gvsbuild build --build-dir d:\gstreamer\gvsbuild_test --enable-gi pycairo  pygobject```

Changed my `PATH` to have 
```
C:\Users\Tim\AppData\Local\Programs\Python\Python39
C:\Users\Tim\AppData\Local\Programs\Python\Python39\Scripts
```
instead and doing the the same thing now but to 
```gvsbuild build --build-dir d:\gstreamer\gvsbuild_testpy39 --enable-gi pycairo  pygobject```
 now, I know I'm probably rebuilding everything, but it's a clean build.
(Compiling cairo tests now).

**Comment by tjwilli58** [comment]
Same thing.
```
meson.build:15:15: ERROR: python3 not found

A full log can be found at D:\gstreamer\gvsbuild_testpy39\build\x64\release\pycairo\_gvsbuild-meson\meson-logs\meson-log.txt
```

**Comment by tjwilli58** [comment]
[gvsbuild-log.txt](https://github.com/wingtk/gvsbuild/files/11713172/gvsbuild-log.txt)

**Comment by tjwilli58** [comment]
[meson-log.txt](https://github.com/wingtk/gvsbuild/files/11713176/meson-log.txt)

**Comment by tjwilli58** [comment]
If I type `python3` in a command prompt, I get the Windows Store wanting me to install Python311 app. Since I have Python39 in my `PATH`, I install the Python3.9 app. That didn't help the meson build find `python3`.

**Comment by tjwilli58** [comment]
meson_options.txt has this:
`option('python', type : 'string', value : 'python3')`

Maybe change to `value:'python'` ?

**Comment by tjwilli58** [comment]
Edited meson_options.txt file. Now I get

```meson.build:15:15: ERROR: python not found```

**Comment by danyeaw** [comment]
How did you install Python, where is it installed to, and what is your $env:Path?

**Comment by tjwilli58** [comment]
I used the Windows installer. 
```
where python
C:\Users\Tim\AppData\Local\Programs\Python\Python39\python.exe
C:\Users\Tim\AppData\Local\Microsoft\WindowsApps\python.exe
```
Typing `python` from a command prompt, I get 3.9.3, but just `py` I get 3.11.3.
```
C:\Program Files\Microsoft Visual Studio\2022\Community>py -0
 -V:3.11 *        Python 3.11 (64-bit)
 -V:3.9           Python 3.9 (64-bit)
 -V:3.8           Python 3.8 (64-bit)
 -V:3.7           Python 3.7 (64-bit)
 -V:ContinuumAnalytics/Anaconda39-64 Anaconda 2022.10
```
I did notice that the gvsbuild was installed using 3.11:
```
meson.build:15:15: ERROR: python3 not found

A full log can be found at D:\gstreamer\gvsbuild_testpy39\build\x64\release\pycairo\_gvsbuild-meson\meson-logs\meson-log.txt
WARNING: Running the setup command as `meson [options]` instead of `meson setup [options]` is ambiguous and deprecated.
Traceback (most recent call last):
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\Lib\site-packages\gvsbuild\utils\builder.py", line 525, in build
    if self.__build_one(p):
       ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\Lib\site-packages\gvsbuild\utils\builder.py", line 661, in __build_one
    skip_deps = proj.build()
                ^^^^^^^^^^^^
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\Lib\site-packages\gvsbuild\projects\pycairo.py", line 38, in build
    Meson.build(self)
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\Lib\site-packages\gvsbuild\utils\base_builders.py", line 49, in build
    self._setup_meson_and_ninja(ninja_build, meson_params, add_path)
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\Lib\site-packages\gvsbuild\utils\base_builders.py", line 81, in _setup_meson_and_ninja
    self.exec_vs(cmd, add_path=add_path)
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\Lib\site-packages\gvsbuild\utils\base_project.py", line 195, in exec_vs
    self.builder.exec_vs(
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\Lib\site-packages\gvsbuild\utils\builder.py", line 935, in exec_vs
    self.__execute(
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\Lib\site-packages\gvsbuild\utils\builder.py", line 1037, in __execute
    subprocess.check_call(args, cwd=working_dir, env=env, shell=True)
  File "C:\Users\Tim\AppData\Local\Programs\Python\Python311\Lib\subprocess.py", line 413, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'C:\Users\Tim\.local\pipx\venvs\gvsbuild\Scripts\python.exe d:\gstreamer\gvsbuild_testpy39\tools\meson-1.1.0\meson.py d:\gstreamer\gvsbuild_testpy39\build\x64\release\pycairo d:\gstreamer\gvsbuild_testpy39\build\x64\release\pycairo\_gvsbuild-meson --prefix d:\gstreamer\gvsbuild_testpy39\gtk\x64\release --buildtype debugoptimized' returned non-zero exit status 1.
Error: pycairo build failed
```
I need to uninstall that and pip install the other version. That doesn't explain why python3 isn't found when I had 3.11 in my PATH though.

**Comment by tjwilli58** [comment]
I have anaconda installed, but it's not in my PATH. This was installed back in November, and I installed gstreamer successfully in December. I may have done this in a virtual enviroment? I don't remember, but this could be my issue.
I reinstalled gvsbuild under my Python 3.9 pip with
```
python -m pip install pipx
pipx install gvsbuild --force 
```

I still get the error about not finding python3, but at the end of the traceback, The anaconda subprocess..py is being found:

```
meson.build:15:15: ERROR: python3 not found

A full log can be found at D:\gstreamer\gvsbuild_testpy39\build\x64\release\pycairo\_gvsbuild-meson\meson-logs\meson-log.txt
WARNING: Running the setup command as `meson [options]` instead of `meson setup [options]` is ambiguous and deprecated.
Traceback (most recent call last):
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\lib\site-packages\gvsbuild\utils\builder.py", line 525, in build
    if self.__build_one(p):
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\lib\site-packages\gvsbuild\utils\builder.py", line 661, in __build_one
    skip_deps = proj.build()
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\lib\site-packages\gvsbuild\projects\pycairo.py", line 38, in build
    Meson.build(self)
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\lib\site-packages\gvsbuild\utils\base_builders.py", line 49, in build
    self._setup_meson_and_ninja(ninja_build, meson_params, add_path)
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\lib\site-packages\gvsbuild\utils\base_builders.py", line 81, in _setup_meson_and_ninja
    self.exec_vs(cmd, add_path=add_path)
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\lib\site-packages\gvsbuild\utils\base_project.py", line 195, in exec_vs
    self.builder.exec_vs(
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\lib\site-packages\gvsbuild\utils\builder.py", line 935, in exec_vs
    self.__execute(
  File "C:\Users\Tim\.local\pipx\venvs\gvsbuild\lib\site-packages\gvsbuild\utils\builder.py", line 1037, in __execute
    subprocess.check_call(args, cwd=working_dir, env=env, shell=True)
  File "d:\anaconda3\lib\subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'C:\Users\Tim\.local\pipx\venvs\gvsbuild\Scripts\python.exe d:\gstreamer\gvsbuild_testpy39\tools\meson-1.1.0\meson.py d:\gstreamer\gvsbuild_testpy39\build\x64\release\pycairo d:\gstreamer\gvsbuild_testpy39\build\x64\release\pycairo\_gvsbuild-meson --prefix d:\gstreamer\gvsbuild_testpy39\gtk\x64\release --buildtype debugoptimized' returned non-zero exit status 1.
Error: pycairo build failed
```

```
  File "d:\anaconda3\lib\subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
```
I don't understand why that version of `subprocess.py` is being found.

**Comment by tjwilli58** [comment]
Created a virtual env and tried to build pycairo and pygobject. Using commit 9db9d4bfba9a31a9057ca51c652f7fbee5b154d8 (5/3/23). Same problem.
```
meson.build:15:15: ERROR: python3 not found

A full log can be found at D:\gstreamer\gvsbuild_testpy39\build\x64\release\pycairo\_gvsbuild-meson\meson-logs\meson-log.txt
WARNING: Running the setup command as `meson [options]` instead of `meson setup [options]` is ambiguous and deprecated.
Traceback (most recent call last):
  File "D:\GitClones\gvsbuild\.venv\lib\site-packages\gvsbuild\utils\builder.py", line 525, in build
    if self.__build_one(p):
  File "D:\GitClones\gvsbuild\.venv\lib\site-packages\gvsbuild\utils\builder.py", line 661, in __build_one
    skip_deps = proj.build()
  File "D:\GitClones\gvsbuild\.venv\lib\site-packages\gvsbuild\projects\pycairo.py", line 38, in build
    Meson.build(self)
  File "D:\GitClones\gvsbuild\.venv\lib\site-packages\gvsbuild\utils\base_builders.py", line 49, in build
    self._setup_meson_and_ninja(ninja_build, meson_params, add_path)
  File "D:\GitClones\gvsbuild\.venv\lib\site-packages\gvsbuild\utils\base_builders.py", line 81, in _setup_meson_and_ninja
    self.exec_vs(cmd, add_path=add_path)
  File "D:\GitClones\gvsbuild\.venv\lib\site-packages\gvsbuild\utils\base_project.py", line 195, in exec_vs
    self.builder.exec_vs(
  File "D:\GitClones\gvsbuild\.venv\lib\site-packages\gvsbuild\utils\builder.py", line 935, in exec_vs
    self.__execute(
  File "D:\GitClones\gvsbuild\.venv\lib\site-packages\gvsbuild\utils\builder.py", line 1037, in __execute
    subprocess.check_call(args, cwd=working_dir, env=env, shell=True)
  File "C:\Users\Tim\AppData\Local\Programs\Python\Python39\lib\subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'D:\GitClones\gvsbuild\.venv\Scripts\python.exe d:\gstreamer\gvsbuild_testpy39\tools\meson-1.1.0\meson.py d:\gstreamer\gvsbuild_testpy39\build\x64\release\pycairo d:\gstreamer\gvsbuild_testpy39\build\x64\release\pycairo\_gvsbuild-meson --prefix d:\gstreamer\gvsbuild_testpy39\gtk\x64\release --buildtype debugoptimized' returned non-zero exit status 1.
Error: pycairo build failed
```
At least the correct subprocess.py was called. I'll pull the latest and try again.

**Comment by tjwilli58** [comment]
```
(.venv) d:\GitClones\gvsbuild>pip uninstall gvsbuild
Found existing installation: gvsbuild 2023.4.0
Uninstalling gvsbuild-2023.4.0:
  Would remove:
    d:\gitclones\gvsbuild\.venv\lib\site-packages\gvsbuild-2023.4.0.dist-info\*
    d:\gitclones\gvsbuild\.venv\lib\site-packages\gvsbuild\*
    d:\gitclones\gvsbuild\.venv\scripts\gvsbuild.exe
Proceed (Y/n)? y
  Successfully uninstalled gvsbuild-2023.4.0
```

```
Successfully installed gvsbuild-2023.6.0
WARNING: You are using pip version 22.0.4; however, version 23.1.2 is available.
You should consider upgrading via the 'D:\GitClones\gvsbuild\.venv\Scripts\python.exe -m pip install --upgrade pip' command.

(.venv) d:\GitClones\gvsbuild>pip list
Package           Version
----------------- -----------
build             0.10.0
click             8.1.3
colorama          0.4.6
gvsbuild          2023.6.0
markdown-it-py    2.2.0
mdurl             0.1.2
packaging         23.1
pip               22.0.4
Pygments          2.15.1
pyproject_hooks   1.0.0
rich              13.4.1
setuptools        58.1.0
shellingham       1.5.0.post1
tomli             2.0.1
typer             0.9.0
typing_extensions 4.6.3
```
Same thing. I'm going to see what happens if I check out a commit from December. 
Author:		Dan Yeaw <dan@yeaw.me>
Date:		5 months ago (12/28/2022 11:23:13 AM)
Committer:	GitHub <noreply@github.com>
Commit hash:	79804bb5892e4e4e3561c1599d375e2190d2c4ee
Child:		6ac122f9
Parents:		03d67976 d8310ddb

Merge pull request #797 from wingtk/gst-plugin-dependencies

Fix gst-python missing base plugins dependency
Contained in branches:
main

Contained in tags:
2023.1.0
2023.1.1
2023.2.0
2023.3.0
2023.4.0
2023.5.0
2023.6.0

Derives from tag: 2022.6.0 + 37 commits

**Comment by tjwilli58** [comment]
```
(.venv) d:\GitClones\gvsbuild>pip show gvsbuild
Name: gvsbuild
Version: 2022.6.0
Summary: GTK stack for Windows
Home-page:
Author: Ignacio Casal Quinteiro
Author-email: qignacio@amazon.com
License: GPL-2.0-only
Location: d:\gitclones\gvsbuild\.venv\lib\site-packages
Requires: build, typer
Required-by:
```

**Comment by tjwilli58** [comment]
Still can't find python3.

**Comment by danyeaw** [comment]
Hi @tjwilli58 If you are able to get to a minimal reproducible issue, could you please open a new issue for this?

It sounds to me that there is an issue with your environment. Have you tried to create a minimal meson project that depends on python? That would help determine if this is an issue with gvsbuild.

**Comment by tjwilli58** [comment]
@danyeaw ,
I'll give it a shot. I've never actually created a meson project, just following instructions to build something someone else wrote.
I'm starting to think it's not gvsbuild, since when I checked out older versions of the repos, I get the same problem. It seems that the only project that I have a problem with is pycairo. I just finished building gst-plugins-good w/o a problem. 

The virtual python environment does find the correct python. I've edited my PATH to only have the Windows python launcher in  creating the venv. If I type "python" in a command line, the python I used to create the venv is found.

Is there a way to `pip install pycairo` so gvsbuild doesn't try to build it?
